File: webhooks/python/samples/folder.moved.py
# Sensitive Folder Move Alert + Owner Notification

from flask import Flask, request, abort
import hmac
import hashlib
import requests
import os

app = Flask(__name__)

# Keywords to flag sensitive folders
SENSITIVE_KEYWORDS = ["contract", "finance", "nda"]

# Generate HMAC-SHA256 signature
def get_secret_hash(secret, payload):
    return "sha256=" + hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()

@app.route("/webhook", methods=["POST"])
def webhook():
    signature = request.headers.get("x-docspace-signature-256")
    computed_signature = get_secret_hash(
        os.environ.get("WEBHOOK_SECRET_KEY"),
        request.get_data()
    )

    if signature != computed_signature:
        app.logger.warning("Invalid signature")
        abort(401)

    event = request.json.get("event", {})
    payload = request.json.get("payload", {})

    trigger = event.get("trigger")
    if trigger != "folder.moved":
        return "", 200

    folder_title = payload.get("title")
    root_id = payload.get("rootId")
    created_by = payload.get("createBy")
    folder_id = payload.get("id")

    app.logger.info(f'Folder moved: "{folder_title}"')

    try:
        # Step 1: Get user info
        user_info = requests.get(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/people/{created_by}",
            headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
        ).json()

        user_email = user_info.get("email")
        user_name = f"{user_info.get('firstName', '')} {user_info.get('lastName', '')}".strip()

        # Step 2: Notify user
        if user_email:
            requests.post(
                os.environ.get("NOTIFY_WEBHOOK_URL"),
                json={
                    "to": user_email,
                    "subject": "Folder moved",
                    "message": f'The folder "{folder_title}" has been moved to a new location in the room structure.'
                },
                headers={"Content-Type": "application/json"}
            )
            app.logger.info(f"User notified: {user_email}")

        # Step 3: Get folder details (for tags)
        folder_meta = requests.get(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/files/{folder_id}",
            headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
        ).json()

        folder_tags = folder_meta.get("response", {}).get("tags", [])
        lower_title = folder_title.lower() if folder_title else ""
        combined = " ".join([lower_title] + [tag.lower() for tag in folder_tags])

        is_sensitive = any(keyword in combined for keyword in SENSITIVE_KEYWORDS)

        # Step 4: If sensitive - notify admin
        if is_sensitive:
            requests.post(
                os.environ.get("NOTIFY_WEBHOOK_URL"),
                json={
                    "to": os.environ.get("ADMIN_EMAIL"),
                    "subject": "Sensitive folder moved",
                    "message": f'User "{user_name}" moved a sensitive folder: "{folder_title}"\n\nFolder ID: {folder_id}\nRoom ID: {root_id}'
                },
                headers={"Content-Type": "application/json"}
            )

            app.logger.info("Admin notified about sensitive move")

        return "", 200

    except Exception as err:
        app.logger.error(f"Error handling folder.moved: {err}")
        return "", 500

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 3000))
    app.run(host="0.0.0.0", port=port)

File: webhooks/nodejs/samples/folder.moved.js
// Sensitive Folder Move Alert + Owner Notification

const express = require("express");
const crypto = require("crypto");
const axios = require("axios");
require("dotenv").config();

const app = express();
app.use(express.json({ verify: rawBodySaver }));

function rawBodySaver(req, res, buf) {
  req.rawBody = buf.toString();
}

function getSecretHash(secret, payload) {
  return (
    "sha256=" +
    crypto.createHmac("sha256", secret).update(payload).digest("hex")
  );
}

// Keywords to flag sensitive folders
const SENSITIVE_KEYWORDS = ["contract", "finance", "nda"];

app.post("/webhook", async (req, res) => {
  const signature = req.headers["x-docspace-signature-256"];
  const computedSignature = getSecretHash(process.env.WEBHOOK_SECRET_KEY, req.rawBody);
  if (signature !== computedSignature) return res.sendStatus(401);

  const { trigger } = req.body.event;
  if (trigger !== "folder.moved") return res.sendStatus(200);

  const { title: folderTitle, rootId, createBy, id: folderId } = req.body.payload;

  console.log(`Folder moved: "${folderTitle}"`);

  try {
    // Step 1: Get user info
    const userInfo = await axios.get(
      `${process.env.PORTAL_URL}/api/2.0/people/${createBy}`,
      {
        headers: {
          Authorization: `Bearer ${process.env.API_KEY}`
        }
      }
    );

    const userEmail = userInfo.data?.email;
    const userName = `${userInfo.data?.firstName || ""} ${userInfo.data?.lastName || ""}`.trim();

    // Step 2: Notify user
    if (userEmail) {
      await axios.post(
        process.env.NOTIFY_WEBHOOK_URL,
        {
          to: userEmail,
          subject: "Folder moved",
          message: `The folder "${folderTitle}" has been moved to a new location in the room structure.`
        },
        { headers: { "Content-Type": "application/json" } }
      );
      console.log(`User notified: ${userEmail}`);
    }

    // Step 3: Get folder details (for tags)
    const folderMeta = await axios.get(
      `${process.env.PORTAL_URL}/api/2.0/files/${folderId}`,
      {
        headers: {
          Authorization: `Bearer ${process.env.API_KEY}`
        }
      }
    );

    const folderTags = folderMeta.data?.response?.tags || [];
    const lowerTitle = folderTitle.toLowerCase();
    const combined = [lowerTitle, ...folderTags.map(t => t.toLowerCase())].join(" ");

    const isSensitive = SENSITIVE_KEYWORDS.some(keyword => combined.includes(keyword));

    // Step 4: If sensitive - notify admin
    if (isSensitive) {
      await axios.post(
        process.env.NOTIFY_WEBHOOK_URL,
        {
          to: process.env.ADMIN_EMAIL,
          subject: "Sensitive folder moved",
          message: `User "${userName}" moved a sensitive folder: "${folderTitle}"\n\nFolder ID: ${folderId}\nRoom ID: ${rootId}`
        },
        { headers: { "Content-Type": "application/json" } }
      );

      console.log("Admin notified about sensitive move");
    }

    res.sendStatus(200);
  } catch (err) {
    console.error("Error handling folder.moved:", err.response?.data || err.message);
    res.sendStatus(500);
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`);
});
